File: additionalProblem/week20/soobin/OpenChat.java
package week20.soobin;

import java.util.*;

public class OpenChat {
    private class User {
        String uid, nickname;

        User(String id, String nickname) {
            this.uid = id;
            this.nickname = nickname;
        }

        public String enter() {
            return nickname + "님이 들어왔습니다.";
        }

        public String leave() {
            return nickname + "님이 나갔습니다.";
        }
    }

    private class Log {
        char command;
        String uid;

        Log(char command, String uid) {
            this.command = command;
            this.uid = uid;
        }

        public String printLog() {
            User user = totalUser.get(uid);
            return command == 'E' ? user.enter() : user.leave();
        }
    }

    private Map<String, User> totalUser;

    private List<Log> processRecords(String[] records) {
        List<Log> logs = new ArrayList<>();

        for (String record : records) {
            StringTokenizer st = new StringTokenizer(record);
            String type = st.nextToken();
            String uid = st.nextToken();

            if (type.charAt(0) != 'C') logs.add(new Log(type.charAt(0), uid));
            if (type.charAt(0) == 'L') continue;

            String nickname = st.nextToken();
            User user = new User(uid, nickname);
            totalUser.put(uid, user);
        }

        return logs;
    }

    private List<String> printLogs(List<Log> logs) {
        List<String> answer = new ArrayList<>();
        for (Log log : logs) answer.add(log.printLog());

        return answer;
    }

    public List<String> solution(String[] records) {
        totalUser = new HashMap<>();
        List<Log> logs = processRecords(records);
        return printLogs(logs);
    }

    public static void main(String[] args) {
        OpenChat oc = new OpenChat();
        String[] records = new String[] {"Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234", "Enter uid1234 Prodo", "Change uid4567 Ryan"};
        System.out.println(oc.solution(records));
    }
}

File: additionalProblem/week20/week20_oh/OpenChat.cpp

#include <string>
#include <map>
#include <vector>
using namespace std;

string tmp, uid, name;
map<string, string> m;

void strToken(string& s) {
    tmp = s.substr(s.find(" ") + 1, s.length());
    uid = tmp.substr(0, tmp.find(" "));
    name = tmp.substr(tmp.find(" ") + 1, tmp.length());
}

vector<string> solution(vector<string> record) {
    vector<string> answer;
    
    for (int i = 0; i < record.size(); i++) {
        if (record[i][0] == 'L') continue;
        strToken(record[i]);
        m[uid] = name;
    }
    
    for (int i = 0; i < record.size(); i++) {
        strToken(record[i]);
        if (record[i][0] == 'E')
            answer.push_back(m[uid] + "님이 들어왔습니다.");
        else if (record[i][0] == 'L')
            answer.push_back(m[uid] + "님이 나갔습니다.");
    }
    return answer;
}
